### ISSUE TITLE
0.14.5: "Generic Knightly Order" used as faction in quest dialogue

### ISSUE THREAD

**Issue by JayH2971** [Interkarma/daggerfall-unity#2510]
User reported the following text in a quest:

![image](https://user-images.githubusercontent.com/37541932/233721239-be481476-c1e0-4e3a-8171-e5a806fe9786.png)

Referenced quest is Q0C20Y02. The issue is the selected faction, "Generic Knightly Order."

A similar issue was seen previously regarding Generic Temple: [https://github.com/Interkarma/daggerfall-unity/issues/2368](https://github.com/Interkarma/daggerfall-unity/issues/2368)

EDIT: User adds that this occurred in Daenia, where there is no knightly order to choose from.

**Comment by Interkarma** [comment]
Interesting the quest still found a location to compile. Do you know which specific quest this is? Did user provide a save?

**Comment by JayH2971** [comment]
User sends this save file: 
[SAVE31__Q0C20Y02_Generic_Knightly_Order_quest_text.zip](https://github.com/Interkarma/daggerfall-unity/files/11300102/SAVE31__Q0C20Y02_Generic_Knightly_Order_quest_text.zip)

**Comment by KABoissonneault** [comment]
Reposting my findings while investigating for #2589 

I just ran the test in DOS. I made a save with 50 reputation with the Glenmoril Witches and went to the NPC in front of the Thieves Guild to ask for quests.

I first tried in Daenia, and could never get this quest. I did have a few "gave the quest to some spellsword" type messages, which usually occurs when a quest fails to resolve. Maybe I needed to be in a region with a knight order.

When I tried again in Daggerfall, I did eventually land on it.

![image](https://github.com/Interkarma/daggerfall-unity/assets/5789925/62e991ec-5ceb-4b9a-99b7-798d6a9aaa60)

![image](https://github.com/Interkarma/daggerfall-unity/assets/5789925/d1d53b1e-547c-4238-941d-19239561a6a5)

This is indeed a knightly guard... of the Totambu knight order, not Daggerfall. 

The NPC did look like a guard at the site.

![image](https://github.com/Interkarma/daggerfall-unity/assets/5789925/59656319-243c-46be-86bc-7c9d8fe44188)

Seems to me like the solution to this issue is that Knightly_Guard should resolve to a random knight order, not the local one. This would fix the problem for regions without a knight order.

**Comment by John-Leitch** [comment]
Looks like I was misguided in my first attempt. Definitely interested in fixing this (and possibly contributing more to the project as I come to understand it), so bear with me as I ask some dumb questions.

My first inclination is to find a way to solve this without a hardcoded list of knight order ids. I've stepped through the QRC/QBN parser to familiarize myself with it, and have a relatively good understanding of what's going on there. Looking at adjacent cases in `Person.GetFactionTypeFactionID`, it seems the `FactionFile.FactionTypes.KnightlyGuard` case could return a randomized faction similar to those that return `Person.GetRandomFactionOfType`. However, there's no faction type specific to the knight order, so that solution won't work verbatim.

Looking at `FACTION.TXT`, I'm not really seeing anything distinct that would allow the knight orders to be safely projected into a collection. Am I overlooking something, or is there no easy way around working with a hardcoded collection of knight order ids?

**Comment by KABoissonneault** [comment]
I think hardcoding the knight orders is fine for now. Unlike temples, knight orders are not properly assigned a "knightly order" type in FACTION.TXT.
Whenever a quest requests "Knightly_Guard", `DaggerfallWorkshop.Game.Questing.Person.GetFactionTypeFactionID` should return one of the 10 default knightly orders, instead of 844 as it does right now.

This may affect other quests. For example, K0C30Y03 "A Noble's Debts" has a branch if you have a reputation above 10 with the Knightly_Guard NPC. The fix would probably be a good thing, since it's unlikely that anyone has good rep with Generic Knight Order. Q0C20Y02 is probably the only quest that named the faction of the guard, though.

@JayH2971 Do you ever use Knightly_Guard in your quests? Do you foresee anything bad from changing this?

**Comment by KABoissonneault** [comment]
Does it help if you change the Person declaration in Q0C20Y02.txt to
```
Person _other_ factiontype Knightly_Guard male remote
```
I'm thinking maybe this is a faction type that only supports male NPCs. 

I think we have some gender issues with NPCs still. We've had a big refactor not that long ago from Ferital here: https://github.com/Interkarma/daggerfall-unity/pull/2145. We might have to go and make all the Knightly_Guard persons male.

Honestly I wish I knew how this sprite selection worked, worth investigating

**Comment by John-Leitch** [comment]
Alright, adding the `male` attribute seems to have done it. That, or I'm very lucky. I'll get a PR put together and submit it for review.

**Comment by KABoissonneault** [comment]
> Alright, adding the `male` attribute seems to have done it. That, or I'm very lucky. I'll get a PR put together and submit it for review.

Actually, looking at FACTION.TXT, it seems intended for knights to have a female sprite.
```
		#368
		type: 2
		name: The Knights of the Dragon
		rep: 0
		summon: -1
		region: 18
		power: 20
		flags: 0
		ally: 371
		face: -1
		race: 3
		flat: 334 16
		flat: 182 45
		sgroup: 7
		ggroup: 9
		minf: 5
		maxf: 40
```
Note the two flat fields. `Person.GetFlatDetailsString` shows that male will pick flat1, and female will pick flat2 (or individuals will always pick flat1). 
Flag 182.45 is indeed the one shown in your screenshot.
![image](https://github.com/Interkarma/daggerfall-unity/assets/5789925/ac51c454-6536-43d9-a69f-f99e67998c08)

So I don't think there's any issues with your test, it's just that knight ladies don't have a distinct guard sprite.

**Comment by KABoissonneault** [comment]
Fixed in 8640acd038765ef9e6dea5e750798ba6afec3d0b
